src/crypto/openssl/openssl_wrappers.h

```diff
@@ -47,245 +54,184 @@ namespace crypto
       }
     }
 
-    class Unique_BIO
+    /// Returns the error string from an error code
+    inline std::string error_string(int ec)
     {
-      std::unique_ptr<BIO, void (*)(BIO*)> p;
+      return ERR_error_string((unsigned long)ec, NULL);
+    }
+
+    /*
+     * Unique pointer wrappers for SSL objects, with SSL' specific constructors
+     * and destructors. Some objects need special functionality, others are just
+     * wrappers around the same template interface Unique_SSL_OBJECT.
+     */
+
+    /// Generic template interface for different types of objects below
+    /// If there are no c-tors in the derived class that matches this one,
+    /// pass `nullptr` to the CTOR/DTOR parameters and make sure to implement
+    /// and delete the appropriate c-tors in the derived class.
+    template <class T, T* (*CTOR)(), void (*DTOR)(T*)>
+    class Unique_SSL_OBJECT
+    {
+    protected:
+      /// Pointer owning storage
+      std::unique_ptr<T, void (*)(T*)> p;
 
     public:
-      Unique_BIO() : p(BIO_new(BIO_s_mem()), [](auto x) { BIO_free(x); })
+      /// C-tor with new pointer via T's c-tor
+      Unique_SSL_OBJECT() : p(CTOR(), DTOR)
       {
         CHECKNULL(p.get());
       }
-      Unique_BIO(const void* buf, int len) :
-        p(BIO_new_mem_buf(buf, len), [](auto x) { BIO_free(x); })
+      /// C-tor with pointer created in base class
+      Unique_SSL_OBJECT(T* ptr, void (*dtor)(T*), bool check_null = true) :
+        p(ptr, dtor)
       {
-        CHECKNULL(p.get());
+        if (check_null)
```

Comment:
I don't understand the reason for this branch, shouldn't we always check that the underlying object isn't null?

Reply:
Not always. The cases where I override with `false` were the cases that didn't check in the past.

Special mention to `UniqueX509` which had a comment:
```
// p == nullptr is OK (e.g. wrong format)
```

Note, this is *only* for cases where the derived class provides the pointer, not on the case we construct the object in the templated base class.

Reply:
Right, that's correct, in those cases we do indeed not want that check!